File: src/main/java/com/paragon/client/ui/animation/Animation.java
package com.paragon.client.ui.animation;

import net.minecraft.util.math.MathHelper;

import java.util.function.Supplier;

/**
 * @author Tigermouthbear, linustouchtips, Wolfsurge
 * @since 06/08/2021
 * Easings
 * @since 12/05/2022
 */
public class Animation {

    // animation time
    public Supplier<Float> time;

    // animation current state
    private State currentState = State.STATIC;
    private long currentStateStart = 0;

    // animation previous state
    private State previousState = State.STATIC;
    private boolean initialState;

    // The easing to apply
    private final Supplier<Easing> easing;

    public Animation(Supplier<Float> time, boolean initialState, Supplier<Easing> easing) {
        this.time = time;
        this.initialState = initialState;
        this.easing = easing;

        // start as expanded
        if (initialState) {
            previousState = State.EXPANDING;
        }
    }

    /**
     * Gets the animation length (0 to 1)
     *
     * @return The animation length (0 to 1)
     */
    public double getAnimationFactor() {
        if (currentState.equals(State.EXPANDING)) {
            return easing.get().ease(MathHelper.clamp((System.currentTimeMillis() - currentStateStart) / time.get(), 0, 1));
        } else if (currentState.equals(State.RETRACTING)) {
            return easing.get().ease(MathHelper.clamp(1 - (System.currentTimeMillis() - currentStateStart) / time.get(), 0, 1));
        }

        return previousState.equals(State.EXPANDING) ? 1 : 0;
    }

    /**
     * Gets the initial state
     *
     * @return The initial state
     */
    public boolean getState() {
        return initialState;
    }

    /**
     * Sets the state
     *
     * @param expand Expand or retract
     */
    public void setState(boolean expand) {
        if (expand) {
            currentState = State.EXPANDING;
            initialState = true;
        } else {
            currentState = State.RETRACTING;
        }

        // reset time
        currentStateStart = System.currentTimeMillis();
    }

    public enum State {

        /**
         * Expands the animation
         */
        EXPANDING,

        /**
         * Retracts the animation
         */
        RETRACTING,

        /**
         * No animation
         */
        STATIC
    }
}

File: src/main/java/com/paragon/client/ui/animation/Animation.kt
package com.paragon.client.ui.animation

import net.minecraft.util.math.MathHelper
import java.util.function.Supplier

/**
 * @author Wolfsurge
 * @since 13/07/22
 */
class Animation(val length: Supplier<Float>, private val initialState: Boolean, val easing: Supplier<Easing>) {

    // Time since last state update
    var lastMillis: Long = 0L

    // Current state of the animation
    // True = Expanding / Expanded
    // False = Contracting / Collapsed
    var state: Boolean = initialState
        set(value) {
            val factor = getFactor()

            lastMillis = if (!value) {
                System.currentTimeMillis() - ((1 - factor) * length.get().toLong()).toLong()
            } else {
                System.currentTimeMillis() - (factor * length.get().toLong()).toLong()
            }

            field = value
        }

    /**
     * Gets the animation factor (value between 0 and 1)
     *
     * @return The animation factor
     */
    fun getAnimationFactor(): Double {
        return if (state) {
            easing.get().ease(MathHelper.clamp((System.currentTimeMillis() - lastMillis.toDouble()) / length.get().toDouble(), 0.0, 1.0))
        } else {
            easing.get().ease(MathHelper.clamp(1 - (System.currentTimeMillis() - lastMillis.toDouble()) / length.get().toDouble(), 0.0, 1.0))
        }
    }

    /**
     * Resets the animation to the original constructor parameters
     */
    fun resetToDefault() {
        state = initialState

        val factor = getFactor()

        lastMillis = if (initialState) {
            System.currentTimeMillis() - ((1 - factor) * length.get().toLong()).toLong()
        } else {
            System.currentTimeMillis() - (factor * length.get().toLong()).toLong()
        }
    }

    /**
     * For internal use only!
     */
    private fun getFactor(): Double {
        var x: Double = MathHelper.clamp((System.currentTimeMillis() - lastMillis.toDouble()) / length.get().toDouble(), 0.0, 1.0)

        if (!state) {
            x = MathHelper.clamp(1 - (System.currentTimeMillis() - lastMillis.toDouble()) / length.get().toDouble(), 0.0, 1.0)
        }

        return x
    }

    /**
     * Constructor.
     */
    init {
        val factor = getFactor()

        lastMillis = if (initialState) {
            System.currentTimeMillis() - ((1 - factor) * length.get().toLong()).toLong()
        } else {
            System.currentTimeMillis() - (factor * length.get().toLong()).toLong()
        }
    }

}